MaxText/layers/linears.py

```diff
@@ -273,40 +274,81 @@ class MoeBlock(nn.Module):
+    
+    w0_kernel = self.param(
+        'wi_0',
+        nn.with_logical_partitioning(kernel_init, kernel_axes),
+        (num_experts, base_emb_dim, mlp_dim),
+        self.weight_dtype,
+        kernel_in_axis,
+        kernel_out_axis,
+      )
+    w0_kernel = jnp.asarray(w0_kernel, self.dtype)
+    w1_kernel = self.param(
+        'wi_1',
+        nn.with_logical_partitioning(kernel_init, kernel_axes),
+        (num_experts, base_emb_dim, mlp_dim),
+        self.weight_dtype,
+        kernel_in_axis,
+        kernel_out_axis,
+      )
+    w1_kernel = jnp.asarray(w1_kernel, self.dtype)
+    wo_kernel = self.param(
+        'wo',
+        nn.with_logical_partitioning(kernel_init, wo_kernel_axes),
+        (num_experts, mlp_dim, base_emb_dim),
+        self.weight_dtype,
+        kernel_in_axis,
+        kernel_out_axis,
+      )
+    wo_kernel = jnp.asarray(wo_kernel, self.dtype)
+    return w0_kernel, w1_kernel, wo_kernel
+
   @nn.compact
-  def __call__(self, inputs, deterministic: bool = False):
+  def __call__(self, inputs):
+    cfg = self.config
+    inputs = inputs.astype(cfg.dtype)
     gate_logits = DenseGeneral(
-        self.num_experts,
-        dtype=self.dtype,
-        kernel_init=self.kernel_init,
-        kernel_axes=self.kernel_axes,
-        name="gate",
-        quant=self.quant,
-    )(inputs)
-
-    weights, selected_experts = lax.top_k(gate_logits, self.num_experts_per_tok)
-    weights = jax.nn.softmax(weights.astype(jnp.float32), axis=-1)
-    mlp_lnx = jnp.zeros_like(inputs)
-    weights = weights.astype(self.dtype)
-    mlp_lnx = nn.with_logical_constraint(mlp_lnx, ("activation_batch", "activation_length", "activation_embed"))
-
-    # TODO(ranran): have a better solution to remove the loop here
-    for k in range(self.num_experts):
-      weights_exp = jnp.sum(jnp.multiply(selected_experts == k, weights), axis=-1)
-      mlp_lnx_exp = MlpBlock(
-          intermediate_dim=self.config.mlp_dim,
-          activations=self.config.mlp_activations,
-          intermediate_dropout_rate=self.config.dropout_rate,
-          dtype=self.dtype,
-          weight_dtype=self.weight_dtype,
-          name=f"mlp_{k}",
-          config=self.config,
-      )(inputs, deterministic=deterministic)
-
-      mlp_lnx_exp = nn.with_logical_constraint(mlp_lnx_exp, ("activation_batch", "activation_length", "activation_embed"))
-      mlp_lnx_exp = weights_exp[:, :, None] * mlp_lnx_exp
-      mlp_lnx += mlp_lnx_exp
-
-    return mlp_lnx
+            self.num_experts,
+            dtype=self.dtype,
+            kernel_init=self.kernel_init,
+            kernel_axes=self.kernel_axes,
+            name='gate')(inputs)
+    
+    top_k_weights, top_k_indices = jax.lax.top_k(gate_logits, self.num_experts_per_tok)
+    flattened_top_k_weights = top_k_weights.reshape(-1, self.num_experts_per_tok)
+
+    softmax_probs = jax.nn.softmax(flattened_top_k_weights.astype(jnp.float32), axis=-1).astype(self.weight_dtype)
```

Comment:
Yeah, this matches implementation [here](https://github.com/mistralai/mistral-inference/blob/main/src/mistral_inference/moe.py#L29). Just in case, we changed the default `weight_dtype` in base.yml.